## Chunking:

- Different LLM models have different token limit, we can not process millions of tokens at a time

- when we load a large text, will split it into smaller pieces is called as **Chunks**

- These chunks are then Embedded(numerical representation) and stored in a vector database 

- So that relevant chunks can be retrived 

- Types of Chunking methods

    - **CharcterTextSplitter**

    - **RecursiveCharcterTextSplitter**

    - **TokenTextSplitter**

    - **SemanticChunking**

    - NLTKTextSplitter

    - MarkDownTextSplitter

    - SpacyTextSplitter

    - PythonTextSplitter

    - HTMLTextSplitter

**CharcterTextSplitter**

- Just split based on charcters

In [1]:
from langchain.text_splitter import CharacterTextSplitter

In [3]:
text="""
LangChain is an open-source framework designed to facilitate the development of applications that integrate large language models (LLMs) with external data sources and computational logic.
It provides tools, abstractions, and components to help developers build language model-powered applications beyond simple text generation. 
LangChain focuses on enabling chains of operations, where outputs from an LLM can trigger additional logic, database queries, API calls, or document retrieval steps. 
This chaining approach allows LLMs to perform multi-step reasoning, question-answering, summarization, and decision-making over complex workflows.
 """

In [5]:
Splitter=CharacterTextSplitter()

In [7]:
Splitter.split_text(text)

['LangChain is an open-source framework designed to facilitate the development of applications that integrate large language models (LLMs) with external data sources and computational logic.\nIt provides tools, abstractions, and components to help developers build language model-powered applications beyond simple text generation. \nLangChain focuses on enabling chains of operations, where outputs from an LLM can trigger additional logic, database queries, API calls, or document retrieval steps. \nThis chaining approach allows LLMs to perform multi-step reasoning, question-answering, summarization, and decision-making over complex workflows.']

In [19]:
Splitter=CharacterTextSplitter(separator='\n')
Splitter.split_text(text)

['LangChain is an open-source framework designed to facilitate the development of applications that integrate large language models (LLMs) with external data sources and computational logic.\nIt provides tools, abstractions, and components to help developers build language model-powered applications beyond simple text generation. \nLangChain focuses on enabling chains of operations, where outputs from an LLM can trigger additional logic, database queries, API calls, or document retrieval steps. \nThis chaining approach allows LLMs to perform multi-step reasoning, question-answering, summarization, and decision-making over complex workflows.']

In [21]:
len(Splitter.split_text(text))

1

In [17]:
Splitter=CharacterTextSplitter(separator='')
Splitter.split_text(text)

['LangChain is an open-source framework designed to facilitate the development of applications that integrate large language models (LLMs) with external data sources and computational logic.\nIt provides tools, abstractions, and components to help developers build language model-powered applications beyond simple text generation. \nLangChain focuses on enabling chains of operations, where outputs from an LLM can trigger additional logic, database queries, API calls, or document retrieval steps. \nThis chaining approach allows LLMs to perform multi-step reasoning, question-answering, summarization, and decision-making over complex workflows.']

In [24]:
len('LangChain is an open')

20

- the chunk first preference based on sepeartor only

- below examples for every new line we have more than 20 tokens 

- we provided chunk size of 20 tokens,eventhough becuase of seperater the chunk size is more than 20

- this indicates this method will not take care of sentences and boundaries



In [23]:
Splitter=CharacterTextSplitter(separator='\n',
                              chunk_size=20, # least preference
                              chunk_overlap=2)
chunks=Splitter.split_text(text)
chunks

Created a chunk of size 188, which is longer than the specified 20
Created a chunk of size 140, which is longer than the specified 20
Created a chunk of size 166, which is longer than the specified 20
Created a chunk of size 146, which is longer than the specified 20


['LangChain is an open-source framework designed to facilitate the development of applications that integrate large language models (LLMs) with external data sources and computational logic.',
 'It provides tools, abstractions, and components to help developers build language model-powered applications beyond simple text generation.',
 'LangChain focuses on enabling chains of operations, where outputs from an LLM can trigger additional logic, database queries, API calls, or document retrieval steps.',
 'This chaining approach allows LLMs to perform multi-step reasoning, question-answering, summarization, and decision-making over complex workflows.']

In [35]:
len('LangChain is an open-source framework designed to facilitate the development of applications that integrate large language models (LLMs) with external data sources and computational logic.')

188

In [33]:
Splitter=CharacterTextSplitter(separator='\n',
                              chunk_size=5, # least preference
                              chunk_overlap=2)
chunks=Splitter.split_text('hello how are you.Im good')
chunks

['hello how are you.Im good']

In [37]:
t1='LangChain is an open-source framework designed to facilitate the development of applications that integrate large language models (LLMs) with external data sources and computational logic.'
from nltk.tokenize import word_tokenize
for t1 in Splitter.split_text(text):
    print(len(word_tokenize(t1)))

Created a chunk of size 188, which is longer than the specified 5
Created a chunk of size 140, which is longer than the specified 5
Created a chunk of size 166, which is longer than the specified 5
Created a chunk of size 146, which is longer than the specified 5


28
20
29
20


**chunk size means charcters**

In [39]:
text="""
LangChain is an open-source framework designed to facilitate the development of applications that integrate large language models (LLMs) with external data sources and computational logic.
It provides tools, abstractions, and components to help developers build language model-powered applications beyond simple text generation. 
LangChain focuses on enabling chains of operations, where outputs from an LLM can trigger additional logic, database queries, API calls, or document retrieval steps. 
This chaining approach allows LLMs to perform multi-step reasoning, question-answering, summarization, and decision-making over complex workflows.
 """
Splitter=CharacterTextSplitter(separator='',
                              chunk_size=20,
                              chunk_overlap=2)
chunks=Splitter.split_text(text)
chunks

# 'LangChain is an ope'
    # over lap is 2 charcters
# 'pen-source framework'

['LangChain is an ope',
 'pen-source framework',
 'rk designed to facil',
 'ilitate the developm',
 'pment of application',
 'ons that integrate l',
 'large language mode',
 'dels (LLMs) with ext',
 'xternal data sources',
 'es and computational',
 'al logic.\nIt provide',
 'des tools, abstracti',
 'tions, and component',
 'nts to help develope',
 'pers build language',
 'e model-powered appl',
 'plications beyond si',
 'simple text generati',
 'tion. \nLangChain foc',
 'ocuses on enabling c',
 'chains of operation',
 'ons, where outputs f',
 'from an LLM can tri',
 'rigger additional lo',
 'logic, database quer',
 'eries, API calls, or',
 'or document retrieva',
 'val steps. \nThis cha',
 'haining approach all',
 'llows LLMs to perfor',
 'orm multi-step reaso',
 'soning, question-ans',
 'nswering, summarizat',
 'ation, and decision-',
 'n-making over comple',
 'lex workflows.']

In [53]:
len('LangChain is an ope')

19

In [57]:
len('pen-source framework')

20

In [59]:
text="""
LangChain is an open-source framework designed to facilitate the development of applications that integrate large language models (LLMs) with external data sources and computational logic.
It provides tools, abstractions, and components to help developers build language model-powered applications beyond simple text generation. 
LangChain focuses on enabling chains of operations, where outputs from an LLM can trigger additional logic, database queries, API calls, or document retrieval steps. 
This chaining approach allows LLMs to perform multi-step reasoning, question-answering, summarization, and decision-making over complex workflows.
 """
Splitter=CharacterTextSplitter(separator='\n',
                              chunk_size=50,
                              chunk_overlap=2)
chunks=Splitter.split_text(text)
chunks

Created a chunk of size 188, which is longer than the specified 50
Created a chunk of size 140, which is longer than the specified 50
Created a chunk of size 166, which is longer than the specified 50
Created a chunk of size 146, which is longer than the specified 50


['LangChain is an open-source framework designed to facilitate the development of applications that integrate large language models (LLMs) with external data sources and computational logic.',
 'It provides tools, abstractions, and components to help developers build language model-powered applications beyond simple text generation.',
 'LangChain focuses on enabling chains of operations, where outputs from an LLM can trigger additional logic, database queries, API calls, or document retrieval steps.',
 'This chaining approach allows LLMs to perform multi-step reasoning, question-answering, summarization, and decision-making over complex workflows.']

**why we need overlap**

- we seen it is splitting randomly based on seperator and based on chunk

- the senetenece is breaking

- model sees one chunk at a time and convert into numerical format

- if we add overlap the end of the chunk is repated at the start of text

- Context will be continue

- Better retrieval process

**Recursive Text Splitter**

- Recursive means hiercachy means step by step process

- It first tries split by paragraphs(\n\n).

- If the chunk is still too long it splits by sentence(.)

- If still long then by space or charcter

- Ensure chunks are within the given size but keeps context (using overlap)

- Paragraph ====> Sentence ==== > words

In [32]:
from langchain.text_splitter import RecursiveCharacterTextSplitter
text="""
LangChain is an open-source framework designed to facilitate the development of applications that integrate large language models (LLMs) with external data sources and computational logic.
It provides tools, abstractions, and components to help developers build language model-powered applications beyond simple text generation. 
LangChain focuses on enabling chains of operations, where outputs from an LLM can trigger additional logic, database queries, API calls, or document retrieval steps. 
This chaining approach allows LLMs to perform multi-step reasoning, question-answering, summarization, and decision-making over complex workflows.
 """
Splitter=RecursiveCharacterTextSplitter(separators='',
                              chunk_size=50,
                              chunk_overlap=2)
chunks=Splitter.split_text(text)
chunks

['LangChain is an open-source framework designed to',
 'facilitate the development of applications that',
 'integrate large language models (LLMs) with',
 'external data sources and computational logic.',
 'It provides tools, abstractions, and components',
 'to help developers build language model-powered',
 'applications beyond simple text generation.',
 'LangChain focuses on enabling chains of',
 'operations, where outputs from an LLM can trigger',
 'additional logic, database queries, API calls, or',
 'document retrieval steps.',
 'This chaining approach allows LLMs to perform',
 'multi-step reasoning, question-answering,',
 'summarization, and decision-making over complex',
 'workflows.']

In [41]:
from langchain.text_splitter import RecursiveCharacterTextSplitter

- paragraph

- sentence

- spaces

- empty string

In [4]:
seperators=['\n\n','.'," ",""]
text="""Pragraph one: This is the first paragraph.It has two sentences.
Paragraph two: This is the second paragrpah which is quite long and will exceed our chunk size limit
for demenostration purpose.It has several sentences to ensure splitting.Here another
sentence to make it longer.
Pragraph three: Short."""

In [6]:
text="""Pragraph one: This is the first paragraph.It has two sentences.
Paragraph two: This is the second paragrpah which is quite long and will exceed our chunk size limit
for demenostration purpose.It has several sentences to ensure splitting.Here another
sentence to make it longer.
Pragraph three: Short."""

# Paragraph level '/n/n'
len("Pragraph one: This is the first paragraph.It has two sentences.")

63

In [8]:
len("""Paragraph two: This is the second paragrpah which is quite long and will exceed our chunk size limit
for demenostration purpose.It has several sentences to ensure splitting.Here another
sentence to make it longer.""")

213

In [10]:
len("Pragraph three: Short.")

22

if will it split at paragraph level we have following chunk size

- 63, 213 , 22

- but if we provide chunk size has 50 only

- then paragraph1 and paragph2 are large so furthure splitting will happend

    - that furthure split based on sentences 

**Sentence level slpit understanding**

In [44]:
len("Pragraph one: This is the first paragraph.")

42

In [14]:
len("Pragraph one: This is the first paragraph."),len("It has two sentences.")

(42, 21)

In [18]:
len("""Paragraph two: This is the second paragrpah which is quite long and will exceed our chunk size limit
for demenostration purpose."""),len("""It has several sentences to ensure splitting."""),len("""Here another
sentence to make it longer.""")

(128, 45, 40)

- Now first paragraph does not have any problem , it will divide into two chunks

    - 42,21 and both are less than 50

- Second one splitted into 3 chunks

    - 128, 45 and 40. but first chunk has 128 char it is more than 50

    - So furthure split require

    - furthure splitting means space

**Space level split**

- logically there are so many spaces will available

- it will concetrate till max char to reach 50

In [41]:
len("""Paragraph two: This is the second paragrpah which is quite long and will exceed our chunk size limit
for demenostration purpose.""")

len("Paragraph two: This is the second paragrpah which") #49
len("is quite long and will exceed our chunk size")      #44
len("limit for demenostration purpose.") # 33

33

**final**

In [ ]:
len("Pragraph one: This is the first paragraph.") # 42
len("It has two sentences.") # 21
len("Paragraph two: This is the second paragrpah which") #49
len("is quite long and will exceed our chunk size")      #44
len("limit for demenostration purpose.") # 33
len("""It has several sentences to ensure splitting.""") # 45
len("""Here another sentence to make it longer.""") # 40
len("Pragraph three: Short.") # 22

**chunk size after splitting should be less than or equal to provided chunk size**

In [43]:
from langchain.text_splitter import RecursiveCharacterTextSplitter

seperators=['\n\n','.'," ",""]
text="""Pragraph one: This is the first paragraph.It has two sentences.

Paragraph two: This is the second paragrpah which is quite long and will exceed our chunk size limit
for demenostration purpose.It has several sentences to ensure splitting.Here another
sentence to make it longer.

Pragraph three: Short."""


splitter=RecursiveCharacterTextSplitter(chunk_size=50,
                              chunk_overlap=0,
                              separators=seperators)

chunks=splitter.split_text(text)

In [76]:
chunks

['Pragraph one: This is the first paragraph',
 '.It has two sentences.',
 'Paragraph two: This is the second paragrpah',
 'which is quite long and will exceed our chunk',
 'size limit\nfor demenostration purpose',
 '.It has several sentences to ensure splitting',
 '.Here another\nsentence to make it longer.',
 'Pragraph three: Short.']

In [49]:
for i,c in enumerate(chunks):
    print(f"Chunk {i} has {len(c)} chars")

Chunk 0 has 41 chars
Chunk 1 has 21 chars
Chunk 2 has 45 chars
Chunk 3 has 45 chars
Chunk 4 has 37 chars
Chunk 5 has 45 chars
Chunk 6 has 40 chars
Chunk 7 has 24 chars


In [72]:
from langchain.text_splitter import RecursiveCharacterTextSplitter

seperators=['\n\n']
text="""Pragraph one: This is the first paragraph.It has two sentences.
Paragraph two: This is the second paragrpah which is quite long and will exceed our chunk size limit
for demenostration purpose.It has several sentences to ensure splitting.Here another
sentence to make it longer.
Pragraph three: Short."""


splitter=RecursiveCharacterTextSplitter(chunk_size=50,
                              chunk_overlap=0,
                              separators=seperators)

chunks=splitter.split_text(text)
len(chunks)

1

## TokenTextSplitter

- If we want to chunks based on **tokens** then we use **TokenTextSplitter**

- 'Hello' is one token, but 'Hello' has 5 charcters

- CharcterTextSplitter and RecursiveCharcterText splitter based on charcters only

- But TokenTextSplitter completely based on token only

    - tokens can be words,spl char

**tiktoken**

- Official OpenAI tokenizer

- it converts text to tokens and tokens to text

- Count tokens in text

- Split texts so that it fits within token limits

- Estimate cost

- pip install tiktoken

In [45]:
import tiktoken
encoding=tiktoken.encoding_for_model('gpt-3.5-turbo')

In [46]:
text="I love learning Langchain and AI!"
tokens=encoding.encode(text)
tokens

[40, 3021, 6975, 23272, 8995, 323, 15592, 0]

In [47]:
len(tokens)

8

In [48]:
len(text)

33

In [9]:
encoding.encode('langchain'), encoding.decode([5317])

([5317, 8995], 'lang')

In [11]:
encoding.decode(tokens)

'I love learning Langchain and AI!'

In [ ]:
"I love learning Lnagchain and AI!"=['I','love','learning','langchain','and','AI','!']  # 7
[40, 3021, 6975, 98919, 351, 8995, 323, 15592, 0] # 9

In [11]:
encoding.n_vocab

100277

In [23]:
word="I love learning Langchain and AI!"
t=encoding.encode(word)
for i in t:
    print(i,encoding.decode([i]))

40 I
3021  love
6975  learning
23272  Lang
8995 chain
323  and
15592  AI
0 !


In [17]:
encoding.encode('learning')

[21656]

In [53]:
from langchain.text_splitter import TokenTextSplitter


text="""Pragraph one: This is the first paragraph.It has two sentences.
Paragraph two: This is the second paragrpah which is quite long and will exceed our chunk size limit
for demenostration purpose.It has several sentences to ensure splitting.Here another
sentence to make it longer.
Pragraph three: Short."""


splitter=TokenTextSplitter(chunk_size=15, 
                              chunk_overlap=3)
# each chunk will have 15 tokens 
# each token overlap of 3 tokens

chunks=splitter.split_text(text)
len(chunks)

6

In [59]:
chunks

['Pragraph one: This is the first paragraph.It has two sentences.',
 ' two sentences.\nParagraph two: This is the second paragr',
 ' paragrpah which is quite long and will exceed our chunk size',
 ' our chunk size limit\nfor demenostration purpose.It has several',
 'It has several sentences to ensure splitting.Here another\nsentence to make',
 'ence to make it longer.\nPragraph three: Short.']

In [54]:
len(encoding.encode(chunks[0]))

14

In [55]:
chunks[2]

' paragrpah which is quite long and will exceed our chunk size'

In [57]:
word="demenostration"
t=encoding.encode(word)
print(len(t))
for i in t:
    print(i,encoding.decode([i]))

3
28531 dem
12052 eno
55681 stration


- tokenTextsplitter Boundaries based on token only

- demenostration divides into 3 tokens so there is a chance the chunk can break

In [68]:
from langchain.text_splitter import TokenTextSplitter,RecursiveCharacterTextSplitter
text="""ChatGPT helps developers build powerful AI applications efficiently.
      It provideds tools and APIs that make integraton easy and scalable"""

seperators=['\n\n','.'," ",""]
recursive_splitter=RecursiveCharacterTextSplitter(chunk_size=50,
                              chunk_overlap=0,
                              separators=seperators)

recursive_chunks=recursive_splitter.split_text(text)
len(recursive_chunks)

for i,c in enumerate(recursive_chunks):
    print(f"Chunk {i} : {c} has ===> {len(c)} chars")
######################################################################
print('============Tik token =========================')
token_splitter=TokenTextSplitter(chunk_size=10, 
                              chunk_overlap=2)
# each chunk will have 15 tokens 
# each token overlap of 3 tokens

token_chunks=token_splitter.split_text(text)
len(token_chunks)

for i,c in enumerate(token_chunks):
    print(f"Chunk {i} : {c} has ===> {len(c)} chars")

##########################################################################
enc=tiktoken.encoding_for_model('gpt-3.5-turbo')
#encoding=tiktoken.encoding_for_model('gpt-3.5-turbo')
for i,c in enumerate(token_chunks):
    tokens=enc.encode(c)
    print(f"Chunk {i} has {len(tokens)} tokens : {tokens}")

Chunk 0 : ChatGPT helps developers build powerful AI has ===> 42 chars
Chunk 1 : applications efficiently has ===> 24 chars
Chunk 2 : .
      It provideds tools and APIs that make has ===> 45 chars
Chunk 3 : integraton easy and scalable has ===> 28 chars
============Tik token =========================
Chunk 0 : ChatGPT helps developers build powerful AI applications efficiently has ===> 67 chars
Chunk 1 :  applications efficiently.
      It has ===> 35 chars
Chunk 2 :   It provideds tools and APIs that make integ has ===> 45 chars
Chunk 3 :  make integraton easy and scalable has ===> 34 chars
Chunk 0 has 10 tokens : [16047, 38, 2898, 8779, 13707, 1977, 8147, 15592, 8522, 30820]
Chunk 1 has 5 tokens : [8522, 30820, 627, 415, 1102]
Chunk 2 has 10 tokens : [220, 1102, 3984, 82, 7526, 323, 34456, 430, 1304, 5503]
Chunk 3 has 6 tokens : [1304, 8936, 24444, 4228, 323, 69311]


- RecursiveTextsplitter for normal language documents we can use

    - Natural readabiity

    - chunking boundaries based on senetence or paragraph

    - here we are not worrying token limit

- if we feel the token limits are more important becuase this directly impact to cost

    - so at that time go with token text splitter

    - Chunk boundary is based on tokens then token text splitter

## Semantic Embedding Chunking

- **Semantic Embedding chunking** is a strategy used when preparing text data for Embedding models

- So that long documents can be effectively represented as smaller, meaningful chunks 

- It preserves **Semantic Coherence**

    - Coherence means a logical and consistent connection where all parts of a whole fit together well

- Each chunk remains self contained and meaningful

- It involves 3 main steps

- 1. **Structure-aware Segmentation**

    - Use documents structure (headings, paragraphs , bullet points) as initial boundaries

    - Example: Split by Markdown boundaries

    - Split by Paragraphs or Split by HTML tags etc

- 2. **Semantic similarity-based merging**

    - Embed the sentences or paragraphs first

    - Embed means a vector representation of numbers

    - it will compute similarity metric (e.g. cosine similarity) betwen sentences or paragraphs

    - Ex1: we have senetnce1 and Sentence2 divided

    - then start compute the similarity metric between two senetences

    - Merge consecutive sentences or paragraphs that are semantically close each until the tokne limit will reach

- 3. **Overlap and Context preservation**

    - Use small overlaps betwen adjacent chunks to prserve context so that It will help in rerieval in RAG applications

In [1]:
# there are so many embedding models are available 
# Embedding model: SentenceTransformer 

# pip install sentence-transformers
import sentence_transformers
from sentence_transformers import SentenceTransformer
from sklearn.metrics.pairwise import cosine_similarity
import nltk
from nltk.tokenize import sent_tokenize
import numpy as np

In [3]:
model= SentenceTransformer('all-MiniLM-L6-v2')
model  # 384 

SentenceTransformer(
  (0): Transformer({'max_seq_length': 256, 'do_lower_case': False, 'architecture': 'BertModel'})
  (1): Pooling({'word_embedding_dimension': 384, 'pooling_mode_cls_token': False, 'pooling_mode_mean_tokens': True, 'pooling_mode_max_tokens': False, 'pooling_mode_mean_sqrt_len_tokens': False, 'pooling_mode_weightedmean_tokens': False, 'pooling_mode_lasttoken': False, 'include_prompt': True})
  (2): Normalize()
)

In [5]:
text="""
Coherence means a logical and consistent connection where all parts of a whole fit together well. 
In a general sense, it refers to things like an argument, story, or plan being easy to understand because 
it is logically integrated and consistent. 
In physics, coherence describes waves that have a stable, fixed relationship in their phase and frequency, 
which is essential for creating stable interference patterns. 
"""

In [7]:
sent=sent_tokenize(text)

In [9]:
embeddings=model.encode(sent,convert_to_numpy=True)

E:\Anaconda\Lib\site-packages\torch\nn\modules\module.py:1750: FutureWarning: `encoder_attention_mask` is deprecated and will be removed in version 4.55.0 for `BertSdpaSelfAttention.forward`.
  return forward_call(*args, **kwargs)


In [13]:
len(embeddings)

3

In [15]:
len(embeddings[0])

384

In [17]:
# Relation between embeddings[0] ===> sen1  and embeddings[1] ===> sen2
cosine_similarity([embeddings[0]],[embeddings[1]])

array([[0.44335645]], dtype=float32)

In [19]:
cosine_similarity([embeddings[0]],[embeddings[2]])

array([[0.7643918]], dtype=float32)

In [ ]:
# So Sente1 and sent3 should be merge 
# Semantic chunker method from langchain

**Semantic Chunk using langchain**

- step-1: we will read the embedd model

    - Every LLM has related embedd model availabe

    - OpenAI has its own embedd model (With API key)

    - GeminiAI has its own embedd model (with API key)

    - hugging face transformers embedd model its works offline (with out API key)

- Step-2: Pass that embedd model the splitter i.e. semantic chunk that we will take by langchain

- Step-3: Now pass the text we will get chunks based on embeddings

In [ ]:
# pip install -U langchain-experimental
# pip install sentence-transformers

In [38]:
import warnings
warnings.filterwarnings('ignore')

# You will need to install a library for embeddings, such as sentence-transformers
# pip install sentence-transformers
from langchain_community.embeddings import SentenceTransformerEmbeddings

# Create an instance of the embedding model
embeddings = SentenceTransformerEmbeddings(model_name="all-MiniLM-L6-v2")
embeddings

HuggingFaceEmbeddings(client=SentenceTransformer(
  (0): Transformer({'max_seq_length': 256, 'do_lower_case': False, 'architecture': 'BertModel'})
  (1): Pooling({'word_embedding_dimension': 384, 'pooling_mode_cls_token': False, 'pooling_mode_mean_tokens': True, 'pooling_mode_max_tokens': False, 'pooling_mode_mean_sqrt_len_tokens': False, 'pooling_mode_weightedmean_tokens': False, 'pooling_mode_lasttoken': False, 'include_prompt': True})
  (2): Normalize()
), model_name='all-MiniLM-L6-v2', cache_folder=None, model_kwargs={}, encode_kwargs={}, multi_process=False, show_progress=False)

In [40]:
# Initialize the SemanticChunker
# pip install langchain-experimental
from langchain_experimental.text_splitter import SemanticChunker
semantic_chunker = SemanticChunker(embeddings)
semantic_chunker

- CharcterTextSplitter does not required any model

- RecursiveTextSplotter does not required any model

- TikTokenTextSplitter  internally use GPT model

- SemanticChunker need any embedd model

In [44]:
# Example text to split
text = """
AI has come a long way since 1952, when the first documented success of an AI computer program was written by Christopher Strachey, whose checkers program completed a whole game on the Ferranti Mark I computer at the University of Manchester. Thanks to developments in machine learning and deep learning, IBM’s Deep Blue defeated chess grandmaster Garry Kasparov in 1997, and the company’s IBM Watson won Jeopardy! in 2011.  

Since then, generative AI has spearheaded the latest chapter in AI’s evolution, with OpenAI releasing its first GPT model in 2018. This has culminated in OpenAI developing ChatGPT, leading to a proliferation of tools that can process queries to produce relevant text, audio, images and other types of content.

Other companies have followed suit with competing products of their own, including Google’s Gemini, Anthropic’s Claude and DeepSeek’s R1 and V3 models, which made headlines in early 2025 for approaching parity with competing models at a fraction of the operational cost.

AI has also been used to help sequence RNA for vaccines and model human speech, technologies that rely on model- and algorithm-based machine learning and increasingly focus on perception, reasoning and generalization.
"""

# Split the text
chunks = semantic_chunker.split_text(text)
#print(chunks)
# Print the chunks
for i, chunk in enumerate(chunks):
    print(f"Chunk {i+1}:\n{chunk}\n---")

Chunk 1:

AI has come a long way since 1952, when the first documented success of an AI computer program was written by Christopher Strachey, whose checkers program completed a whole game on the Ferranti Mark I computer at the University of Manchester. Thanks to developments in machine learning and deep learning, IBM’s Deep Blue defeated chess grandmaster Garry Kasparov in 1997, and the company’s IBM Watson won Jeopardy! in 2011.
---
Chunk 2:
Since then, generative AI has spearheaded the latest chapter in AI’s evolution, with OpenAI releasing its first GPT model in 2018. This has culminated in OpenAI developing ChatGPT, leading to a proliferation of tools that can process queries to produce relevant text, audio, images and other types of content. Other companies have followed suit with competing products of their own, including Google’s Gemini, Anthropic’s Claude and DeepSeek’s R1 and V3 models, which made headlines in early 2025 for approaching parity with competing models at a fracti

- SemanticChunker works a bit differently than other text splitters you might be familiar with, like RecursiveCharacterTextSplitter. 

- Instead of splitting based on a fixed token limit or character count, SemanticChunker splits text based on the semantic similarity of the sentences.

- There isn't a direct chunk_size or chunk_overlap parameter like in other text splitters. 

- The splitting behavior is primarily controlled by the breakpoint_threshold_type and breakpoint_threshold parameters during initialization. 

- These parameters determine how dissimilar the semantic embeddings of sentences need to be to create a split.

- A lower breakpoint_threshold will result in more frequent splits and therefore smaller chunks, while a higher threshold will result in fewer splits and larger chunks.

- So, while you don't control the token limit directly, you can indirectly influence the size of the chunks by adjusting the breakpoint_threshold.

In [25]:
# Split the text
chunks = semantic_chunker.split_text(text)

# Print the chunks
for i, chunk in enumerate(chunks):
    print(f"Chunk {i+1}:\n{chunk}\n---")

Chunk 1:

AI has come a long way since 1952, when the first documented success of an AI computer program was written by Christopher Strachey, whose checkers program completed a whole game on the Ferranti Mark I computer at the University of Manchester. Thanks to developments in machine learning and deep learning, IBM’s Deep Blue defeated chess grandmaster Garry Kasparov in 1997, and the company’s IBM Watson won Jeopardy! in 2011.
---
Chunk 2:
Since then, generative AI has spearheaded the latest chapter in AI’s evolution, with OpenAI releasing its first GPT model in 2018. This has culminated in OpenAI developing ChatGPT, leading to a proliferation of tools that can process queries to produce relevant text, audio, images and other types of content. Other companies have followed suit with competing products of their own, including Google’s Gemini, Anthropic’s Claude and DeepSeek’s R1 and V3 models, which made headlines in early 2025 for approaching parity with competing models at a fracti

**Problem**

- This chunks divided based on semantic meaning of sentences 

- One chunk means this all the data has a meaning 

- But different chunks has different sizes

- We are not able to control the tokens in each chunk

- Some times we might face the issue with our context window or model max token length might be reached

- then you need to take these chunks and apply **Recursive Text Splitter** 

- we can control the token limit

In [3]:
from langchain import text_splitter

In [5]:
dir(text_splitter)

['CharacterTextSplitter',
 'ElementType',
 'HTMLHeaderTextSplitter',
 'HeaderType',
 'KonlpyTextSplitter',
 'Language',
 'LatexTextSplitter',
 'LineType',
 'MarkdownHeaderTextSplitter',
 'MarkdownTextSplitter',
 'NLTKTextSplitter',
 'PythonCodeTextSplitter',
 'RecursiveCharacterTextSplitter',
 'RecursiveJsonSplitter',
 'SentenceTransformersTokenTextSplitter',
 'SpacyTextSplitter',
 'TextSplitter',
 'TokenTextSplitter',
 'Tokenizer',
 '__all__',
 '__builtins__',
 '__cached__',
 '__doc__',
 '__file__',
 '__loader__',
 '__name__',
 '__package__',
 '__spec__',
 'split_text_on_tokens']